File: fully-written-solutions/leetcode/python/two_sum.py
from typing import List


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]: # type: ignore
        prevMap = {}

        for i, n in enumerate(nums):
            diff = target - n
            if diff in prevMap:
                return [prevMap[diff], i]
            prevMap[n] = i

                
def test_two_sum():
    solution = Solution()
    test_cases = [
        ([2, 7, 11, 15], 9, [0, 1]),
        ([], 0, None),
        ([1], 1, None),
        ([0, 1, 2, 3], 0, None),
        ([-1, -2, -3, -4], -7, [2, 3]),
        ([3, 3], 6, [0, 1])
    ]

    for i, (nums, target, expected_result) in enumerate(test_cases):
        assert solution.twoSum(nums, target) == expected_result
        print(f"Test case {i+1} passed with expected result {expected_result}")

test_two_sum()

File: fully-written-solutions/Sources/leetcode/swift/twosum.swift
/*
1. Two Sum
Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

You can return the answer in any order.

Example 1:

Input: nums = [2,7,11,15], target = 9
Output: [0,1]
Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].
Example 2:

Input: nums = [3,2,4], target = 6
Output: [1,2]
Example 3:

Input: nums = [3,3], target = 6
Output: [0,1]
 
Constraints:

2 <= nums.length <= 104
-109 <= nums[i] <= 109
-109 <= target <= 109
Only one valid answer exists.
*/

class Solution {
    func twoSum(_ nums: [Int], _ target: Int) -> [Int] {
        var map: [Int: Int] = [:]

        for i in 0..<nums.count {
            if let val = map[target - nums[i]] {
                return [val, i]
            } else {
                map[nums[i]] = i
            }
        }

        return [-1, -1]
    }
}

func test() {
    let sol = Solution()
    let testCases = [
        (([2,5,7,9], 9), [0, 2]),
        (([2,3,4], 7), [1, 2])
    ]
    
    for (i, testCase) in testCases.enumerated() {
        let (input, expected) = testCase
        let (arr, target) = input
        let result = sol.twoSum(arr, target)
        assert(result == expected, "Test case \(i + 1) failed. Expected \(expected), got \(result)")
        print("Test case \(i + 1) passed.")
    }
}

test()